inithome/src/rubin/nublado/inithome/provisioner.py

```diff
@@ -0,0 +1,103 @@
+"""Provisioner for user home directories."""
+
+import asyncio
+import logging
+from os import chown, environ, getenv, umask
+from pathlib import Path
+
+MAX_ID = 2**32 - 1  # True for Linux, which is where we run inithome
+
+
+class Provisioner:
+    """Object to oversee idempotent provisioning of home directories."""
+
+    def __init__(self, uid: int, gid: int, homedir: Path) -> None:
+        self._validate(uid, gid)
+        self.uid = uid
+        self.gid = gid
+        self.homedir = homedir
+
+    def _validate(self, uid: int, gid: int) -> None:
+        for item in (uid, gid):
+            if item == 0:
+                raise ValueError("Will not provision for UID/GID 0")
+            if item < 0:
+                raise ValueError("UID/GID must be positive")
+            if item > MAX_ID:
+                raise ValueError(f"UID/GID must be <= {MAX_ID}")
+
+    async def provision(self) -> None:
+        """Provision user home directory.
+
+        This is the only public method.  Given an initialized
+        object with UID, GID, and homedir path, if the path does not
+        exist, create it and set it to mode 0700 for the path
+        components created.
+
+        If the path already exists, verify that it is a directory and owned
+        by the right UID and GID; warn if mode is not correct, but don't treat
+        that as a fatal error.
+        """
+        if self.homedir.exists():
+            if not self.homedir.is_dir():
+                raise RuntimeError(
+                    f"{self.homedir} exists but is not a directory"
+                )
+            # Check ownership and permissions
+            stat_results = self.homedir.stat()
+            if stat_results.st_uid != self.uid:
+                raise RuntimeError(
+                    f"{self.homedir} is owned by {stat_results.st_uid}, not "
+                    f"{self.uid}"
+                )
+            if stat_results.st_gid != self.gid:
+                raise RuntimeError(
+                    f"{self.homedir} is owned by group "
+                    f"{stat_results.st_gid}, not {self.gid}"
+                )
```

Comment:
I'm tempted to just fix the permissions rather than throw a fatal error, although I suppose that doesn't make sense if there are files. Maybe check if the directory is empty, and if so, just fix the permissions?

Specifically, I'm worried about the case where someone runs Nublado without an init container first without realizing it's needed, Kubernetes creates the directory owned as root, and now the user is stuck and needs manual intervention. It would be nice if we could self-heal from that scenario if it won't cause other problems.

Reply:
Done.